### PR TITLE
Secure filedrop

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
- [x] Tests written, or not not needed

### CODE REVIEW

File: app/src/main/java/com/owncloud/android/utils/EncryptionUtils.java [hunk 1]
```diff
@@ -205,16 +239,45 @@ public static DecryptedFolderMetadata decryptFolderMetaData(EncryptedFolderMetad
...
+                                                                       new TypeToken<DecryptedFolderMetadata.Data>() {
+                                                                       }));
 
             files.put(key, decryptedFile);
+
+            // remove from filedrop
+            fileDrop.remove(key);
         }
 
         return decryptedFolderMetadata;
```
Comment (AlvaroBrey) [inline app/src/main/java/com/owncloud/android/utils/EncryptionUtils.java:283]:
```suggestion
        if (fileDrop != null) {
            for (Map.Entry<String, EncryptedFolderMetadata.EncryptedFile> entry : fileDrop.entrySet()) {
                String key = entry.getKey();
                EncryptedFolderMetadata.EncryptedFile encryptedFile = entry.getValue();

                DecryptedFolderMetadata.DecryptedFile decryptedFile = new DecryptedFolderMetadata.DecryptedFile();
                decryptedFile.setInitializationVector(encryptedFile.getInitializationVector());
                decryptedFile.setMetadataKey(encryptedFile.getMetadataKey());
                decryptedFile.setAuthenticationTag(encryptedFile.getAuthenticationTag());

                // decrypt
                String dataJson = EncryptionUtils.decryptStringAsymmetric(encryptedFile.getEncrypted(), privateKey);

                decryptedFile.setEncrypted(EncryptionUtils.deserializeJSON(dataJson,
                                                                           new TypeToken<DecryptedFolderMetadata.Data>() {
                                                                           }));

                files.put(key, decryptedFile);

                // remove from filedrop
                fileDrop.remove(key);
            }
        }

        return decryptedFolderMetadata;
```

**Review by AlvaroBrey — APPROVED** [review]
Some suggestions but nothing blocking.

File: app/src/main/java/com/owncloud/android/datamodel/EncryptedFolderMetadata.java [hunk 2]
```diff
@@ -53,24 +64,25 @@ public void setFiles(Map<String, EncryptedFile> files) {
 
     public static class EncryptedFile {
         private String encrypted;
+        @SerializedName(value = "initializationVector", alternate = "initialisationVector")
```
[the remark this answers is not in the document]
Comment (tobiasKaminsky) [reply app/src/main/java/com/owncloud/android/datamodel/EncryptedFolderMetadata.java:67]: This was a typo in https://github.com/nextcloud/end_to_end_encryption/pull/367
It seems now to be fixed, thus we can remove it.

File: app/src/main/java/com/owncloud/android/ui/adapter/OCFileListDelegate.kt [hunk 3]
```diff
@@ -237,7 +238,9 @@ class OCFileListDelegate(
         bindGridMetadataViews(file, gridViewHolder)
 
         // shares
-        val shouldHideShare = gridView || hideItemOptions || file.isFolder && !file.canReshare() || file.isEncrypted ||
+        val shouldHideShare = gridView || hideItemOptions || file.isFolder && !file.canReshare() ||
+            !file.isFolder && file.isEncrypted ||
+            file.isEncrypted && !EncryptionUtils.supportsSecureFiledrop(file, user) ||
             searchType == SearchType.FAVORITE_SEARCH
```
[the remark this answers is not in the document]
Comment (tobiasKaminsky) [reply app/src/main/java/com/owncloud/android/ui/adapter/OCFileListDelegate.kt:244]: I put every condition in a new line, "separated" by ||.

**Pull request merged**